modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -107,24 +120,30 @@ public void onTopologyLeap(LogicalTopologySnapshot newTopology) {
         }
     };
 
+    /** The logical topology on the last watch event. */
+    private Set<String> logicalTopology = Collections.emptySet();
+
+    /** Watch listener id to unregister the watch listener on {@link DistributionZoneManager#stop()}. */
+    private Long watchListenerId;
```

Comment:
Same as above, this variable should be thread-safe. It's possible to you will try to set and read it from different threads.

Reply:
fixed